parsers/yaml/yaml.cc

```diff
@@ -0,0 +1,2776 @@
+             << (T(TagPrefix)[TagPrefix] * T(TagHandle)[TagHandle] * End)) >>
+          [](Match& _) { return Seq << _[TagPrefix] << _[TagHandle]; },
+
+        In(Tag) *
+            (T(TagGroup)
+             << (T(TagPrefix)[TagPrefix] *
+                 T(ShorthandTag, VerbatimTag)[TagName] * End)) >>
+          [](Match& _) { return Seq << _[TagPrefix] << _[TagName]; },
+
+        In(Tag) * (T(TagGroup) << (T(TagPrefix)[TagPrefix] * End)) >>
+          [](Match& _) { return Seq << _[TagPrefix] << (NonSpecificTag ^ ""); },
+
+        // errors
+
+        In(StreamGroup) * (DirectiveToken[Value] * End) >>
+          [](Match& _) {
+            return err(_(Value), "Directive by itself with no document");
+          },
+
+        In(DocumentGroup) *
+            (T(MaybeDirective)[MaybeDirective] * ~T(NewLine) *
+             End)([](auto& n) {
+              Node dir = n.first[0];
+              Node doc = dir->parent()->parent()->shared_from_this();
+              Node stream = doc->parent()->shared_from_this();
+              return stream->find(doc) < stream->end() - 1;
+            }) >>
+          [](Match& _) {
+            return err(
+              _(MaybeDirective), "Directive without document end marker");
+          },
+
+        In(DocumentGroup, FlowGroup) *
+            T(TagPrefix, ShorthandTag, VerbatimTag)[Tag] >>
+          [](Match& _) { return err(_(Tag), "Invalid tag"); },
+
+        In(DocumentGroup) *
+            (DirectiveToken / T(Document, TagHandle, Stream))[Value] >>
+          [](Match& _) { return err(_(Value), "Syntax error"); },
+      }};
+
+    pass.pre(Document, [](Node n) {
+      n->insert(n->begin(), Directives);
+      return 0;
+    });
+
+    pass.post(Stream, [](Node n) {
+      Node directives = Directives
+        << (TagDirective << (TagPrefix ^ "!") << (TagHandle ^ "!"))
+        << (TagDirective << (TagPrefix ^ "!!")
+                         << (TagHandle ^ "tag:yaml.org,2002:"));
+      n->insert(n->begin(), directives);
+      if (n->size() == 1)
+      {
+        n->push_back(Documents ^ "");
+      }
+      return 0;
+    });
+
+    pass.post(Tag, [](Node n) {
+      if (n->size() == 1)
+      {
+        n->push_back(NonSpecificTag ^ "");
+      }
+      return 0;
+    });
+
+    pass.post([](Node n) {
+      return invalid_tokens(
+        n,
+        {{StreamGroup, "Invalid stream"},
+         {TagDirectiveGroup, "Invalid tag directive"},
+         {TagGroup, "Invalid tag"}});
+    });
+
+    return pass;
+  }
+
+  PassDef flow()
+  {
+    PassDef flow = {
+      "flow",
+      wf_flow,
+      dir::bottomup,
+      {
+        In(FlowMapping, FlowSequence) * T(FlowGroup)[FlowGroup] >>
+          [](Match& _) { return Seq << *_[FlowGroup]; },
+
+        In(FlowSequence) *
+            (T(Value)[Value] * T(Comma, FlowSequenceEnd))([](auto& n) {
+              Location loc = n.first[0]->location();
+              return loc.view() == "-";
+            }) >>
```

Comment:
I think the regex version of T might be useful here:
```suggestion
            (T(Value, "-")[Value] * T(Comma, FlowSequenceEnd)) >>
```